packages/longhorn/package.yaml

```diff
@@ -0,0 +1,5 @@
+url: https://github.com/longhorn/charts.git
+subdirectory: charts/longhorn
+type: git
+commit: 0a88eaae4d7bdce762af942ea490d72aee338bb1
+packageVersion: 00
```

Comment:
@cbron @vincent99 does Longhorn need to split into CRD charts? In https://github.com/rancher/charts/pull/526#issuecomment-673114537 we decided to not split it into a separate CRD chart, but just wanted to confirm that that's still the case.

Reply:
We do need to split it into CRD chart, as we now don't want to remove CRDs on chart uninstall. That was new development from last week. So if you uninstall the longhorn chart it removes everything but CRDs/CRs, and you also have to uninstall the longhorn-crd chart to remove the CRDs/CRs.

Reply:
Hmm so @shuo-wu you'll need to add the following to the `package.yaml`:
```yaml
generateCRDChart:
  enabled: true
```

You'll also need to split the [`crds.yaml`](https://github.com/longhorn/charts/blob/master/charts/longhorn/templates/crds.yaml) file into separate files per CRD via the patch and relocate those files to the `crd/` directory for the scripts to pick it up correctly.

Unfortunately, support for 1) pointing to a combined CRD file and 2) a set of CRDs outside the `crd/` directory is not supported but tracked in https://github.com/rancher/rancher/issues/28320.

Reply:
I am a little bit confused about the longhorn-crd chart uninstallation. Right now Longhorn uses a separate job to clean up the workloads as well as all CRs. And this uninstall job is launched during the longhorn chart uninstallation. If we split the longhorn crd/cr chart uninstallation from the longhorn chart uninstallation, how can we guarantee all of the CRs can be removed during the longhorn crd chart uninstallation? By launching the job again?

Reply:
@shuo-wu is that uninstall job just part of this helm chart or is it wider job thats always used with longhorn ? 

Ideally all charts in the rancher/charts repo follows the same pattern so we can make assumptions based on that. But if this uninstall script is integral to how longhorn works then we might have to make an exception. The typical behavior is that we can uninstall the main chart which leaves behind the CRs and CRDs, and leave it up to user to delete those by uninstalling the CRD chart.

Reply:
@cbron I am not sure what do you mean by `wider job`

This job is responsible for:
  1. Delete the workloads dynamically deployed by Longhorn system. (instead of YAML files)
  2. Cleaning up CRs.

If I remove the uninstaller job from the main/feature chart, those dynamically deployed cannot be removed. And deleting the crd chart somehow cannot clean up all CRDs (may due to the finalizers of the related CRs)
If I retain the uninstaller job in the main chart, the uninstaller will get stuck there when it tries to handle CRs.

Right now we need to either modify the Longhorn uninstaller (to make it clean up extra workloads only) or make an exception for the charts like Longhorn.

Reply:
Sounds like we need to lean towards making an exception, since everyone with a current longhorn chart installed would be affected. I will ask.

Reply:
We talked about this in the architecture meeting and would like to split the CRDs into 2nd chart, and also move the uninstall script to the CRD chart.

Reply:
Do you mean creating a separate CRD chart manually? Then how about deleting the workloads dynamically deployed by Longhorn system?

Reply:
@shuo-wu yes, I believe @cbron is referring to taking an approach similar to [`fleet`](https://github.com/rancher/fleet/tree/master/charts), which has both a `fleet` chart and a `fleet-crd` chart. 

Ideally, the behavior of the CRD chart should be to be able to have sole control over 1) installing CRDs 2) upgrading CRDs 3) removing CRDs. The main chart should not be able to do anything CRD-related, but should still be able to do anything CR / workload related (i.e. creating or cleaning up CRs and workloads, which is part of what the current uninstall script does).

w.r.t. which of the two charts should be responsible for deleting the workloads dynamically, is there a way for the uninstaller chart to be given a flag to only delete the workloads and not clean up the CRDs?

If this is possible, the main chart should be patched to pass in this flag to the uninstall job so that it no longer is doing anything CRD related and you can likely use the logic described [here](https://github.com/rancher/charts/pull/546#discussion_r489052974) in order to modify the longhorn chart to be able to be split into 2 charts.

If this is not possible, I think the best move would be to move the uninstall script to the CRD chart as Caleb recommended. This does mean that the CRD chart for Longhorn specifically will have one additional responsibility not found in any other CRD charts (i.e. it also cleans up workloads and CRs in addition to the CRDs on uninstall), but I think that was called out in the architecture meeting based on [Caleb's comment](https://github.com/rancher/charts/pull/546#discussion_r491224165).

Reply:
The Longhorn uninstaller is responsible for removing the CRs (objects) rather than CRDs. If the CRs can be cleaned up during the main chart uninstallation, the Longhorn uninstaller should be fine.

Reply:
We talked about this with Sheng and since the uninstaller has to run while the operator is still installed in the cluster, we can't move it to the CRD chart or it would have no effect. So we'll have to leave in the primary chart. But we do still want to split the CRD chart out, and that chart would only control the CRDs not any CRs.

Reply:
Sure, then it seems that I need to create a new PR with a new PR name.

What's the difference between manually splitting the CRD chart and using `catalog.cattle.io/provides-gvr: longhorn.io/v1beta1` generate the hidden CRD chart?

Reply:
If the CRD's were in the CRD folder you can split them automatically with package.yaml using `generateCRDChart:  enabled: true` and that will handle all annotations. You can move them with a patch to the CRD directory, the patch is applied in `make prepare` before the CRD split so that should work. 

You can also split them manually. On the CRD chart you'll need to provide all the primary required annotations and the `hidden` annotation, and on the primary chart the `auto-install: ${original_chart_name}-crd=match`.

`provides-gvr` should be set manually on the primary chart either way, it doesn't relate to the crd chart, but instead allows another chart that needs longhorn to function to require it via the UI. Thats unlikely right now but its a standard we like to set.